feat: Display tool calls in dashboard chat

skazinka opened the pull request:
Also includes the tool calls in the context passed back to the agents.

skazinka committed:
feat: Display tool calls in dashboard chat

Also includes the tool calls in the context passed back to the agents.

skazinka commented:
Seems to be getting a number of 504s for steps involving installing k8s, I'll investigate after giving it some time since it's probably an external issue

dwmkerr committed:
Merge branch 'main' into feat/tool-calls-in-chat

skazinka committed:
Satisfy prettier

Seems to have updated over the break

skazinka committed:
Merge branch 'main' into feat/tool-calls-in-chat

skazinka committed:
Fix issue when agent with tools doesn't use them

Agent seems to have problems with an empty array of tool calls, so we delete the key entirely. Otherwise we end up sending the iterator, which fails due to pydantic.

Also tweaks floating-chat to do fewer unnecessary length calculations.

skazinka committed:
const was right after all

skazinka committed:
Put tool calls before message instead of after

Mimics Claude behavior and is also a lot easier to see if there are many tool calls

skazinka committed:
Merge branch 'main' into feat/tool-calls-in-chat

skazinka committed:
Add test for tool call handling in ark-api

skazinka committed:
Merge branch 'feat/tool-calls-in-chat' of https://github.com/mckinsey/agents-at-scale-ark into feat/tool-calls-in-chat

lucaromagnoli commented on services/ark-api/ark-api/src/ark_api/api/v1/openai.py:
```diff
@@ -213,6 +213,16 @@ async def chat_completions(request: ChatCompletionRequest) -> ChatCompletion:
         metadata["annotations"][STREAMING_ENABLED_ANNOTATION] = "true"
 
     try:
+        # Convert messages to dictionaries to avoid Pydantic serialization issues
```
I don't quite get how this is converting Pydantic models to dictionaries. Besides, what serialization issues are you having? You can define a custom serializer. It's usually more efficient to let Pydantic deal with serialisation.

skazinka replied:
There's a bug in Pydantic which means that `messages["tool_calls"]` ends up returning a ValidatorIterator, https://github.com/pydantic/pydantic/issues/9541 The list comprehension is necessary to get the actual tool calls out of the iterator, if there are any. You're probably right that it's a confusing comment though

dwmkerr commented on services/ark-api/ark-api/src/ark_api/api/v1/openai.py:
```diff
@@ -221,9 +231,10 @@ async def chat_completions(request: ChatCompletionRequest) -> ChatCompletion:
             query_spec_dict["conversationId"] = conversation_id
         if timeout:
             query_spec_dict["timeout"] = timeout
+        logger.info(f"query_spec_dict: {query_spec_dict}")
```
probably don't need these just debug messages?

skazinka replied:
Removed

skazinka committed:
PR feedback

- Rewords confusing comment. Hopefully at some point this bug in Pydantic will be fixed.

- Logging removed
